### ISSUE TITLE
how to assign 2nd arg of center

### ISSUE THREAD

**Issue by henrykey** [ujmp/universal-java-matrix-package#7]
Hi, I try your SparseMatrix, and I don't know what data should be put in for method of center for SparseMastrix? My dataset is 18,000 rows with 220,000 columns (features). do the dataset need to be transposed first? my code like this:

```
    DataMatrix mtx = new DataMatrix();//my own code to init;
    SparseMatrix m1 = mtx.svm2SparseMatrix(LoadedData); //to load data
    m1.center(Ret.ORIG, ?, true);// try to use the method to get center and weigh data
```

thanks

Henry

**Comment by holger-arndt** [comment]
The second parameter defines the dimension along which the mean values are calculated. You will probably want to calculate the mean of the rows so use Matrix.ROW or 0.

**Comment by henrykey** [comment]
Got it, thanks.
when I loading little bit more data, it crashed because out of memory. here is some information
INFO: Matrix:80912x3491
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
    at org.ujmp.core.doublematrix.impl.DefaultDenseDoubleMatrix2D.<init>(DefaultDenseDoubleMatrix2D.java:64)
    at org.ujmp.core.doublematrix.factory.DefaultDenseDoubleMatrix2DFactory.zeros(DefaultDenseDoubleMatrix2DFactory.java:34)
    at org.ujmp.core.doublematrix.calculation.AbstractDoubleCalculation.calcNew(AbstractDoubleCalculation.java:52)
    at org.ujmp.core.calculation.AbstractCalculation.calc(AbstractCalculation.java:99)
    at org.ujmp.core.AbstractMatrix.center(AbstractMatrix.java:555)

my code is like this:
        SparseMatrix m2 = (SparseMatrix) m1.center(Ret.NEW, m1.ROW, true);
what should i do for big dataset?

thanks

**Comment by henrykey** [comment]
Now, it works with Ret.LINK. Dose it mean the matrix being  mapped into a file? So, the limited memory still works. I find it return immediately with the option of Ret.LINK. How can I know it finished?

Thanks

**Comment by holger-arndt** [comment]
Link means that the calculation is not executed at once. So it might crash with the same error as soon as you try to retrieve a value from the matrix. Center is not optimized for sparse matrices yet, that's why it's crashing.

**Comment by henrykey** [comment]
Now, I'm try small dataset for SVD now. My code is like this:
        SparseMatrix m = this.svm2SparseMatrix(spMap);
        org.ujmp.core.Matrix m1 = m.center(Ret.NEW, 0, false);
        svd = m1.svd();

I found it take quite long time to finish. here is log msg:
INFO: Matrix:50x5547
Feb 13, 2016 9:27:24 PM jkmmseg.mmsegTest testPCA
INFO: time for PCA: 93 s

Is it normal? How can I improve the performance?

Henry

**Comment by holger-arndt** [comment]
Have you included EJML or Ojalgo in your classpath? Then these libraries will be used for the calculation of SVD and it should be much faster...

**Comment by henrykey** [comment]
I tried Ojalgo. Yes, it's much faster. But I got some issues here. When I first invoked Ojalgo jar into class path, it's doesn't work with code like this
...
Matrix m1 = LoadMyData();
Matrix m0 = m1.center(Ret.NEW, 0, false);
MySVD = m0.svd();
...
I have to change code of 2nd line to: 
Matrix m0 = m1.center(Ret.ORIG, 0, false);

I want to know why? and if I want to invoke specify matrix lib, how can i do it? 

Thanks.
